### PR TITLE
Return an interface when registering metric callbacks.

### PR DESCRIPTION
**Pull request by anuraaga**
This gives us flexibility to provide functionality on a registered observable metric in the future, notably the ability to unregister them. I think it's likely we will add it after some spec discussion since it is good practice to allow any registered stateful thing to be unregistered - it seems idiomatic compared to `ScheduledFuture` returned from `ScheduledExecutorService` for example.

Helps with #4013 

@jkwatson I didn't quite follow the idea in https://github.com/open-telemetry/opentelemetry-java/issues/4013#issuecomment-998955236 but let me know if you want to try a different pattern. Return value, similar to `ScheduledFuture` seems best to me right now.

### CODE REVIEW

**Comment by anuraaga** [comment]
Also one point to note is that this is compile-compatible, any existing apps will still compile without change when updating from 1.10.0-RC1 to 1.10.0-RC2 / stable. It's not binary-compatible so to avoid adding another method in the future, we would need to do this now. I expect there is enough of a chance of wanting to add APIs here to do this proactively, especially given it's not invasive for existing users in any way.

**Comment by jsuereth** [comment]
My $.02 is that you should return a unique type per-instrument.   If we want this to future proof method on async instruments, this doesn't provide enough flexibility as-is.

Note: Originally the metrics API returned instruments on async registration.  I removed it, to make room for structural changes internally.  While I still think we need to make those changes, I don't think we have a hard dependency on returned Instrument types in Async like we did previously, so this should be fine.

**Comment by anuraaga** [comment]
> If we want this to future proof method on async instruments, this doesn't provide enough flexibility as-is.

We'd be able to add a type parameter in the future as in my suggestion, but it doesn't have to be Observable*Measurement and could be instrument-specific. So I think it does have the flexibility, but such an approach may make using that future functionality a bit more awkward. I'd probably lean towards it if we think instrument-specific logic is not likely, or specific return types if we think it's likely.

**Comment by anuraaga** [comment]
Well - I guess the specific return types don't have a significant issue either let me go with that, will update

**Comment by anuraaga** [comment]
Updated

**Comment by jsuereth** [comment]
> We'd be able to add a type parameter in the future as in my suggestion, but it doesn't have to be Observable*Measurement and could be instrument-specific. So I think it does have the flexibility, but such an approach may make using that future functionality a bit more awkward. I'd probably lean towards it if we think instrument-specific logic is not likely, or specific return types if we think it's likely.

I'm not really sure this leads to the best API.  We don't really know if we have a uniform API per-async instrument but we KNOW they differ in important ways (at least Gauge vs. Counter), and we know MOST APIs we're adapting have diverging methods per-instrument, so it's more likely we diverge than stay the same.

Again, we're talking unknown unknowns, but I think it's far better w/ per-instrument async return value (AsynchronousGauge) vs. a single "Observable".

**Review by jsuereth — APPROVED** [review]

**Review by kubawach — APPROVED** [review]
Looks fine :)

**Review by jkwatson — APPROVED** [review]
Do we need another RC release with this change, do you think?

**Comment by anuraaga** [comment]
Yeah planning on doing one today

File: api/all/src/main/java/io/opentelemetry/api/metrics/ObservableLongUpDownCounter.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.api.metrics;
+
+import java.util.function.Consumer;
+
+/**
+ * A reference to an observable metric registered with {@link
+ * DoubleGaugeBuilder#buildWithCallback(Consumer)}.
```
Comment (jack-berg) [inline api/all/src/main/java/io/opentelemetry/api/metrics/ObservableLongUpDownCounter.java:12]: Typo here.

File: api/all/src/main/java/io/opentelemetry/api/metrics/ObservableDoubleUpDownCounter.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.api.metrics;
+
+import java.util.function.Consumer;
+
+/**
+ * A reference to an observable metric registered with {@link
+ * DoubleGaugeBuilder#buildWithCallback(Consumer)}.
```
Comment (jack-berg) [inline api/all/src/main/java/io/opentelemetry/api/metrics/ObservableDoubleUpDownCounter.java:12]: Typo here.

**Review by jack-berg — APPROVED** [review]

**Comment by mateuszrzeszutek** [comment]
Nice 👍 
Thanks!

**Pull request merged**